### PR TITLE
fix: require all search terms to match in resource search

### PR DESCRIPTION
**Pull request by adriangohjw**
## Problem

Studio resource search treated each word in a query as an independent match — a page matching any single term was returned. Searching `"apple banana"` could surface pages titled `"apple pie"` or `"banana bread"`, which made results noisy and harder to narrow down.

## Solution

**Breaking Changes**

- [x] Yes - this PR contains breaking changes
  - Multi-word searches now require every term to match (AND semantics). Users who relied on broad OR-style results may see fewer matches for the same query.
- [ ] No - this PR is backwards compatible

**Improvements**:

- Changed `getSearchResults` to combine search terms with `AND` instead of `OR`, so all words in the query must appear in the resource title.
- Extracted a shared `tokenizeSearchQuery` helper in `apps/studio/src/schemas/resource.ts` that trims, lowercases, splits on whitespace, and deduplicates terms. Both the schema layer and `resource.service.ts` use this helper for consistent term processing.

**Bug Fixes**:

- None

## Before & After Screenshots

N/A — no visual UI changes.

## Tests

**Manual Verification Steps**:

- [ ] Open Studio search (⌘K / Ctrl+K) on a site with several pages.
- [ ] Search for a single word (e.g. `"budget"`) and confirm matching pages are returned as before.
- [ ] Search for two words where only one appears in a page title (e.g. `"apple banana"` when a page is titled `"apple pie"`) and confirm that page is **not** returned.
- [ ] Search for two words that both appear in a page title (e.g. `"annual budget"` for a page titled `"Annual Budget Report"`) and confirm the page **is** returned.
- [ ] Search with mixed casing (e.g. `"ANNUAL budget"`) and confirm case-insensitive matching still works.

**New scripts**:

- None

**New dependencies**:

- None

**New dev dependencies**:

- None

<details><summary><h3>Greptile Summary</h3></summary>

This PR changes Studio's resource search from OR semantics (any term matches) to AND semantics (all terms must match), so a query like `"apple banana"` no longer surfaces pages that only contain one of the two words. It also extracts a shared `tokenizeSearchQuery` helper into `resource.utils.ts` for consistent lowercasing, whitespace splitting, and deduplication.

- `resource.service.ts`: `eb.or(...)` replaced with `eb.and(...)` so every token in the query must prefix-match a word boundary in the resource title.
- `resource.utils.ts` (new): `tokenizeSearchQuery` trims, lowercases, splits on `/\s+/`, filters empty strings, and deduplicates via `Set`; unit tested in the accompanying `resource.utils.test.ts`.
- Router tests updated to reflect the new semantics — partial-match pages are now expected to be excluded, with new test cases for mid-title matches, case-insensitivity, and prefix boundary behavior.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the OR→AND change is well-scoped, the router's early-return guard prevents the empty-terms edge case from reaching the database layer, and the new tests cover the key behavioral boundaries.

The core logic change (one operator swap) is small and directly tested. tokenizeSearchQuery is independently unit-tested for all edge cases. The only non-critical issue is that the multi-term ordering block now computes a constant score for every returned row and can be simplified or removed, but this does not affect correctness.

resource.service.ts — the ordering block (lines 1362–1388) is now effectively dead code and the accompanying comment is stale; worth cleaning up in a follow-up.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/studio/src/server/modules/resource/resource.service.ts | Changed OR → AND in the WHERE clause so all search terms must match; ordering block is now dead code (all results always score the same constant sum, so effective order is lastUpdatedAt only) |
| apps/studio/src/server/modules/resource/resource.utils.ts | New file exporting tokenizeSearchQuery — trims, lowercases, splits on whitespace, filters empty strings, and deduplicates via Set; implementation is correct and well-tested |
| apps/studio/src/server/modules/resource/__tests__/resource.utils.test.ts | New unit tests for tokenizeSearchQuery covering empty string, whitespace-only, multiple spaces, casing, deduplication, and leading/trailing whitespace — all edge cases addressed |
| apps/studio/src/server/modules/resource/__tests__/resource.router.test.ts | Router tests updated to reflect AND semantics: OR-matching test cases removed, replaced with tests confirming partial-match exclusion, mid-title matching, case-insensitivity, and word-boundary prefix matching |
| apps/studio/src/components/Searchbar/SearchModal.tsx | Cosmetic change only — blank line added before the component export; no functional changes |

</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[User types search query] --> B[useSearchQuery / debounce 300ms]
    B --> C[tRPC resource.search]
    C --> D{query empty?}
    D -- yes --> E[Return recentlyEdited only]
    D -- no --> F[tokenizeSearchQuery\ntrim · lowercase · split · dedupe]
    F --> G[getSearchResults]
    G --> H[WHERE AND each term\nilike 'term%' OR '% term%']
    H --> I{searchTerms > 1?}
    I -- yes --> J[ORDER BY sum of term.length scores DESC\n always constant — all terms matched]
    I -- no --> K[Skip ordering step]
    J --> L[ORDER BY lastUpdatedAt DESC]
    K --> L
    L --> M[Return paginated results]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[User types search query] --> B[useSearchQuery / debounce 300ms]
    B --> C[tRPC resource.search]
    C --> D{query empty?}
    D -- yes --> E[Return recentlyEdited only]
    D -- no --> F[tokenizeSearchQuery\ntrim · lowercase · split · dedupe]
    F --> G[getSearchResults]
    G --> H[WHERE AND each term\nilike 'term%' OR '% term%']
    H --> I{searchTerms > 1?}
    I -- yes --> J[ORDER BY sum of term.length scores DESC\n always constant — all terms matched]
    I -- no --> K[Skip ordering step]
    J --> L[ORDER BY lastUpdatedAt DESC]
    K --> L
    L --> M[Return paginated results]
```

</a>
</details>

<sub>Reviews (5): Last reviewed commit: ["test: update resource.router tests to re..."](https://github.com/opengovsg/isomer/commit/a9cba76cfb6480062e263bb43cd5c7a88955f883) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39746685)</sub>

### CODE REVIEW

File: apps/studio/src/schemas/resource.ts [hunk 1]
```diff
@@ -113,10 +113,17 @@ export interface ResourceItemContent {
...
+export const MAX_SEARCH_QUERY_LENGTH = 256
+
 export const searchSchema = z
   .object({
     siteId: z.string(),
-    query: z.string().trim().optional(),
+    query: z
+      .string()
+      .trim()
+      .max(MAX_SEARCH_QUERY_LENGTH, "Search query is too long")
```
Comment (dcshzj) [inline apps/studio/src/schemas/resource.ts]: question: is this feedback surfaced to the user?

Comment (mvincentong) [reply apps/studio/src/schemas/resource.ts]: I do not think this message is surfaced today. `useSearchQuery` only reads `data` / `isLoading` from `trpc.resource.search.useInfiniteQuery` and drops the query `error`, so a >256-char search rejects at the tRPC input layer but both consumers fall back to empty results: global search renders `NoResultsState`, and `ResourceSelector` renders its zero-result state. If we want users to see `Search query is too long`, the hook probably needs to expose `isError` / `error.message` and the two consumers need an error state, or the input should be capped before calling the query.

Comment (adriangohjw) [reply apps/studio/src/schemas/resource.ts]: @dcshzj updated it to
1. increase query limit (from 256 to 512, which should handle p99 of search query)
2. display 2 new message - one for query exceeding 512 characters which will not show results, while another one for if its more than 10 words which still show results but warn/tip the user to reduce the number of words. I  would need to check with Sehyun regarding the design and copywriting

File: apps/studio/src/schemas/resource.ts [hunk 2]
```diff
@@ -113,10 +113,22 @@ export interface ResourceItemContent {
   parentId: string | null
 }
 
+// Prevent DoS via unbounded query length and excessive ILIKE conditions;
+// value is arbitrary - adjust if needed
+export const MAX_SEARCH_QUERY_LENGTH = 512
+export const MAX_SEARCH_TERMS = 10
+
+export const tokenizeSearchQuery = (query: string): string[] =>
+  query.trim().toLowerCase().split(/\s+/).filter(Boolean)
```
[the remark this answers is not in the document]
Comment (adriangohjw) [reply apps/studio/src/schemas/resource.ts]: done

File: apps/studio/src/schemas/resource.ts [hunk 3]
```diff
@@ -113,10 +113,22 @@ export interface ResourceItemContent {
   parentId: string | null
 }
 
+// Prevent DoS via unbounded query length and excessive ILIKE conditions;
+// value is arbitrary - adjust if needed
+export const MAX_SEARCH_QUERY_LENGTH = 512
```
[the remark this answers is not in the document]
Comment (adriangohjw) [reply apps/studio/src/schemas/resource.ts]: nah i lazy

File: apps/studio/src/components/Searchbar/SearchModal.tsx [hunk 4]
```diff
@@ -42,6 +49,15 @@ export const SearchModal = ({ siteId, isOpen, onClose }: SearchModalProps) => {
     }, []),
   })
 
+  const { isQueryTooLong, hasExcessiveWords } = useMemo(
+    () => ({
+      isQueryTooLong: searchValue.length > MAX_SEARCH_QUERY_LENGTH,
+      hasExcessiveWords:
+        tokenizeSearchQuery(searchValue).length > MAX_SEARCH_TERMS,
```
[the remark this answers is not in the document]
Comment (adriangohjw) [reply apps/studio/src/components/Searchbar/SearchModal.tsx]: done

**Review by seaerchin — APPROVED** [review]
